### PR TITLE
[Offline Transformations] Reshape Layer Pruning Transformation Support

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Details:
Valid reshape operation pruning mask propagation support. 

Reshape mask propagator checks input and output dimensions from the beginning till first shape differences or min sizes of in/out shapes and propagate mask only for dimensions which are completely the same from the beginning of the shapes.
Such way to propagate masks through reshape keeps callback connections in valid state

### Tickets:
 - 67418

TODO:

- [x] Tests
- [x] Beautiful illustration

### CODE REVIEW

**Comment by daniil-lyakhov** [comment]
[Actual diff](https://github.com/openvinotoolkit/openvino/pull/9350/commits/8bd17789ef7f8d3a9b3cd374aa812beb2408f3f7)

**Comment by daniil-lyakhov** [comment]
Ready for review, will be opened after #8926 will be merged

**Comment by daniil-lyakhov** [comment]
Reshape mask is connecting in following way:
image
Connections `[weights mask -> output mask]` and `[weights mask -> output mask]` are removing channels from dimensions which change their size (as it described in the pr description). Other connection just copying value. 

To prevent wrong dimension propagation in case zeros are propagated "down", like in #8926, step with `initialize_dependencies` was added

**Comment by daniil-lyakhov** [comment]
Should fix and cover by a tests following topologies 

- [ ] Mobilenet_v2 ATSS
- [ ] Mobilenet_v2 SSD

**Comment by daniil-lyakhov** [comment]
Performance build: 1581

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 1]
```diff
@@ -582,19 +592,180 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
...
+        auto pooling_by_reduce = pattern::wrap_type<opset6::ReduceMin, opset6::ReduceMax, opset6::ReduceMean>({inputs, weights});
+
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+            const auto & pattern_map = m.get_pattern_value_map();
+            const auto m_weights = pattern_map.at(weights);
+            const auto & m_input = pattern_map.at(inputs);
+            const auto & m_output = pattern_map.at(pooling_by_reduce);
+
+            const auto constant = std::dynamic_pointer_cast<opset6::Constant>(m_weights.get_node_shared_ptr());
+            if (!constant) return false;
```
Comment (GlebKazantaev) [inline src/common/offline_transformations/src/pruning/propagate_masks.cpp:609]: Move to pattern.

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 2]
```diff
@@ -582,19 +592,180 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
...
+
+            // Check reduce operation reduces only dimension without masks
+            if (auto input_mask = getMask(m_input)) {
+                auto output_mask = std::make_shared<Mask>(m_output.get_partial_shape().rank().get_length());
+                const auto reduce_dims = constant->cast_vector<int>();
```
Comment (GlebKazantaev) [inline src/common/offline_transformations/src/pruning/propagate_masks.cpp:614]:
```suggestion
                const auto reduce_dims = constant->cast_vector<int64_t>();
```

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 3]
```diff
@@ -582,19 +592,180 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
...
+            const auto & m_output = pattern_map.at(reshape);
+
+            // Check if this reshape is before group convolution
+            // In such case this reshape should be processed by GroupConvolutionReshape pass
+            for (const auto inp : m_output.get_target_inputs())
+                if (std::dynamic_pointer_cast<opset6::GroupConvolution>(inp.get_node()->shared_from_this()))
+                    return false;
+
+            const auto constant = std::dynamic_pointer_cast<opset6::Constant>(m_weights.get_node_shared_ptr());
+            if (!constant) return false;
```
Comment (GlebKazantaev) [inline src/common/offline_transformations/src/pruning/propagate_masks.cpp:665]: move to pattern

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 4]
```diff
@@ -582,19 +592,180 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
...
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+            const auto & pattern_map = m.get_pattern_value_map();
+            const auto m_weights = pattern_map.at(weights);
+            const auto & m_input = pattern_map.at(inputs);
+            const auto & m_output = pattern_map.at(reshape);
+
+            // Check if this reshape is before group convolution
+            // In such case this reshape should be processed by GroupConvolutionReshape pass
+            for (const auto inp : m_output.get_target_inputs())
+                if (std::dynamic_pointer_cast<opset6::GroupConvolution>(inp.get_node()->shared_from_this()))
```
Comment (GlebKazantaev) [inline src/common/offline_transformations/src/pruning/propagate_masks.cpp:661]:
```suggestion
                if (is_type<opset6::GroupConvolution>(inp.get_node()))
```

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 5]
```diff
@@ -582,19 +592,180 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
...
+
+                auto input_mask_row = input_mask.get();
+                auto output_mask_row = output_mask.get();
+                input_mask->add_callback([output_mask_row](Mask::Ptr cur_mask) -> bool {
+                    cur_mask->copy_value_from_mask(output_mask_row);
+                    return true;
+                }, output_mask);
+                output_mask->add_callback([input_mask_row, reduce_dims](Mask::Ptr cur_mask) -> bool{
+                    // Propagate masks through dimension only if this dimension isn't reduced
+                    for (size_t dim = 0; dim < std::min(cur_mask->size(), input_mask_row->size()); ++dim)
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply src/common/offline_transformations/src/pruning/propagate_masks.cpp:624]: discussed offline.

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp (hunk 4 above)
Comment (daniil-lyakhov) [reply src/common/offline_transformations/src/pruning/propagate_masks.cpp]: Done

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp (hunk 3 above)
Comment (daniil-lyakhov) [reply src/common/offline_transformations/src/pruning/propagate_masks.cpp]: Done

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp (hunk 2 above)
Comment (daniil-lyakhov) [reply src/common/offline_transformations/src/pruning/propagate_masks.cpp]: Done

File: src/common/offline_transformations/include/mask_attribute.hpp [hunk 6]
```diff
@@ -40,10 +40,20 @@ class Mask : public std::vector<std::set<uint64_t>>,
             : std::vector<value_type>(shape.rank().get_length()) {
     }
 
+    explicit Mask(const ngraph::PartialShape & shape, const bool adjust_value)
+            : std::vector<value_type>(shape.rank().get_length()),
+              m_adjust_value(adjust_value) {
+    }
+
```
Comment (daniil-lyakhov) [inline src/common/offline_transformations/include/mask_attribute.hpp:47]: Never used

**Review by alexsu52 — APPROVED** [review]
It looks good to me.

**Comment by daniil-lyakhov** [comment]
What's new:
* Support for `shape_of` branching in groupconv reshape
* Support of calculating values as output shape for reshape operation

Now Efficientnet_b0 int8 + pruning transformed to IR by latest MO are pruned correctly

File: src/common/offline_transformations/include/mask_attribute.hpp [hunk 7]
```diff
@@ -196,8 +201,17 @@ class Mask : public std::vector<std::set<uint64_t>>,
...
     }
 
+    bool adjust_value() const {
+        return m_adjust_value;
+    }
+
 private:
     bool m_is_shape_like{false};
+    // Flag - if constant this mask is related to should be pruned
+    // by value. Means shape of the constant remains the same
```
Comment (GlebKazantaev) [inline src/common/offline_transformations/include/mask_attribute.hpp:211]: smth wrong with this sentence.

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 8]
```diff
@@ -640,6 +646,97 @@ class ngraph::pass::mask_propagation::Reduce : public MatcherPass {
...
+                    return false;
+            }
+
+            // Check reshape operation reshape only dimension without masks
+            if (auto input_mask = getMask(m_input)) {
+                auto output_mask = std::make_shared<Mask>(m_output.get_partial_shape().rank().get_length());
+                auto weights_mask = std::make_shared<Mask>(m_output.get_partial_shape().rank().get_length(), true);
+
+                const auto input_shape = m_input.get_shape();
+                const auto output_shape = constant->cast_vector<size_t>();
```
Comment (GlebKazantaev) [inline src/common/offline_transformations/src/pruning/propagate_masks.cpp:684]: Values in Reshape constant can be negative.
Please cast to int64_t

Comment (daniil-lyakhov) [reply src/common/offline_transformations/src/pruning/propagate_masks.cpp]: Done

File: src/common/offline_transformations/include/mask_attribute.hpp (hunk 7 above)
Comment (daniil-lyakhov) [reply src/common/offline_transformations/include/mask_attribute.hpp:211]: Done, please check

File: src/common/offline_transformations/include/mask_attribute.hpp [hunk 9]
```diff
@@ -196,8 +201,20 @@ class Mask : public std::vector<std::set<uint64_t>>,
         m_need_initialization = true;
     }
 
+    bool adjust_value() const {
+        return m_adjust_value;
+    }
+
 private:
     bool m_is_shape_like{false};
+    // Flag - if a constant this mask is related to should be pruned
```
Comment (GlebKazantaev) [inline src/common/offline_transformations/include/mask_attribute.hpp:210]: `Flag - if a constant this mask is related to should be pruned by value` - please change it.

**Review by GlebKazantaev — APPROVED** [review]

File: src/common/offline_transformations/include/mask_attribute.hpp (hunk 9 above)
Comment (daniil-lyakhov) [reply src/common/offline_transformations/include/mask_attribute.hpp]: Done

**Pull request merged**